packages/demo/backend/scripts/attribution.ts

```diff
@@ -0,0 +1,148 @@
+  transport: http(RPC_URL),
+})
+
+function extract16ByteSuffix(hex: Hex): Hex | null {
+  const raw = hex.slice(2)
+  if (raw.length < 32) return null // need at least 16 bytes
+  const suffix = `0x${raw.slice(-32)}` as Hex // last 16 bytes (32 hex chars)
+  return /^0x[0-9a-fA-F]{32}$/.test(suffix) ? suffix : null
+}
+
+async function main() {
+  const latest = await client.getBlockNumber()
+  const from = latest - BigInt(BLOCKS_BACK)
+  console.log(
+    `Scanning logs ${from} -> ${latest} in ${CHUNK_SIZE}-block chunks for UserOperationEvent at ${ENTRYPOINT} (target suffix ${TARGET_SUFFIX})`,
+  )
+
+  type DecodedUserOp = {
+    sender: Address
+    nonce: bigint
+    initCode: Hex
+    callData: Hex
+    callGasLimit: bigint
+    verificationGasLimit: bigint
+    preVerificationGas: bigint
+    maxFeePerGas: bigint
+    maxPriorityFeePerGas: bigint
+    paymasterAndData: Hex
+    signature: Hex
+  }
+
+  let found = false
+
+  for (
+    let chunkEnd = latest;
+    chunkEnd >= from && !found;
+    chunkEnd -= BigInt(CHUNK_SIZE)
+  ) {
+    const tentativeStart = chunkEnd - BigInt(CHUNK_SIZE - 1)
+    const chunkStart = tentativeStart > from ? tentativeStart : from
+    const userOperationEvent =
+      'event UserOperationEvent(bytes32 indexed userOpHash, address indexed sender, address indexed paymaster, uint256 nonce, bool success, uint256 actualGasCost, uint256 actualGasUsed)'
+
+    const logs = await client.getLogs({
+      address: ENTRYPOINT,
+      event: parseAbiItem(userOperationEvent),
+      fromBlock: chunkStart,
+      toBlock: chunkEnd,
+    })
+
+    // Deduplicate tx hashes (newest-first) and iterate once
+    const seen = new Set<string>()
+    const uniqueTxs: Hex[] = []
+    for (let idx = logs.length - 1; idx >= 0; idx--) {
+      const h = (logs[idx] as any).transactionHash as Hex
+      if (!h) continue
+      if (!seen.has(h)) {
+        seen.add(h)
+        uniqueTxs.push(h)
+      }
+    }
+    console.log(`Found ${uniqueTxs.length} unique txs`)
+
+    for (const txHash of uniqueTxs) {
+      if (found) break
+      const tx = await client.getTransaction({ hash: txHash as any })
+
+      const input = tx.input as Hex
+      let decoded
+      try {
+        decoded = decodeFunctionData({ abi: entryPointAbi, data: input })
+      } catch {
+        continue
+      }
+      if (decoded.functionName !== 'handleOps') continue
+
+      const [ops] = decoded.args as [DecodedUserOp[], Address]
+      for (let i = 0; i < ops.length; i++) {
+        const op = ops[i]
+        const callDataSuffix = extract16ByteSuffix(op.callData)
+        const initCodeSuffix = extract16ByteSuffix(op.initCode)
+        if (
+          callDataSuffix === TARGET_SUFFIX ||
+          initCodeSuffix === TARGET_SUFFIX
+        ) {
+          console.log(
+            `MATCH block=${tx.blockNumber} tx=${tx.hash} opIndex=${i} sender=${op.sender}`,
+          )
+          found = true
+          break
+        }
+      }
+    }
+  }
```

Comment:
🤔 nested for loops get a little hard to read. it might be worth moving the inner loops into helper functions. that would also serve to break this function up a bit